File: solutions/103. Binary Tree Zigzag Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        queue=deque()
        queue.append(root)
        
        res=[]
        left=True
        
        while len(queue)!=0:
            lst=[]
            queue_len=len(queue)
            for i in range(queue_len):
                node=queue.popleft()
                lst.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            if left:
                res.append(lst)
            else:
                res.append(lst[::-1])
            left=not left
        
        return res
            
        
        
        

File: solutions/103. Binary Tree Zigzag Level Order Traversal.java
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        
        List<List<Integer>> res = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        if (root==null) return res;
        boolean invert = true;
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> temp = new ArrayList<>();
            while (size>0) {
                TreeNode node = queue.poll();
                size--;
                temp.add(node.val);
                if (node.left!=null) queue.offer(node.left);
                if (node.right!=null) queue.offer(node.right); 
            }    
            if (invert) {
                res.add(temp);
                invert=false;
            }
            else {
                 Collections.reverse(temp);
                res.add(temp);
                invert=true;
            }
           
        }
        
        return res;
    }
}
